File: String_Algorithms/Finding Borders/sol.py
################################################
#                                              #
#   author: raghav_0901 (Raghav Dalmia)        #
#   updated on: 06-10-2020, 11:44:19 pm IST   #
#                                              #
################################################

from math import *

if __name__ == "__main__":
    s = input()
    n = len(s)
    dp = [0]

    i, j = 0, 1
    while(j < n):
        if(s[i] == s[j]):
            dp.append(i + 1)
            i = i + 1
            j = j + 1
        else:
            if(i == 0):
                dp.append(0)
                j = j + 1
            else:
                i = dp[i - 1]

    i, ans = n - 1, []
    while(i > 0 and dp[i] > 0):
        ans.append(dp[i])
        i = dp[i] - 1

    ans.sort()
    for i in ans:
        print(i, end=" ")

File: String_Algorithms/Finding Borders/sol.cpp
/***********************************************
 *                                             *
 *   author: raghav_0901 (Raghav Dalmia)       *
 *   updated on: 06-10-2020, 11:56:01 pm IST   *
 *                                             *
\***********************************************/

#include <bits/stdc++.h>
using namespace std;

void precompute(int n, string s, vector<int> &dp)
{
    // similar to KMP precompute on pattern string
    int i = 0, j = 1;
    while (j < n)
    {
        if (s[i] == s[j])
        {
            dp[j] = i + 1;
            i++;
            j++;
        }
        else
        {
            if (i == 0)
                j++;
            else
                i = dp[i - 1];
        }
    }
}

int main( )
{
    string s;
    cin >> s;
    int n = s.length( );
    vector<int> dp(n, 0);

    precompute(n, s, dp);

    int i = n - 1;
    vector<int> ans;
    while (i > 0 && dp[i] > 0)
    {
        ans.push_back(dp[i]);
        i = dp[i] - 1;
    }
    reverse(ans.begin( ), ans.end( ));
    for (int i : ans) cout << i << " ";
}
